### DISCUSSION TITLE
ObjCRuntime.RuntimeException: 'Can't register the class {...} when the dynamic registrar has been linked away.'

### DISCUSSION THREAD

**Discussion by jeanplevesque** [unoplatform/uno#11518] [Q&A]
I've recently [updated an Uno class library](https://github.com/nventive/ExtendedSplashScreen/pull/65) and I now get the following error at runtime on iOS.
`ObjCRuntime.RuntimeException: 'Can't register the class Nventive.ExtendedSplashScreen.ExtendedSplashScreen when the dynamic registrar has been linked away.'`

I don't really understand what changed to cause this. The code changes mainly affect the Android platform.
The other change is in the pipelines: we now use hosted agents.
I suspect that the build difference comes from a different default behavior between VS 17.2.5 (previously used in our pipelines) and whatever version is installed on the `windows-2022` images.

**Do we know how to fix this?**

**Comment by jeromelaban** [comment]
There are multiple possible reasons for this, but the main two are:
- The library does not contain a `net6.0-ios` TargetFramework, and the `net6.0-ios` app falls back to netstandard2.0
- The library does have a `net6.0-ios16.1` target framework and app is built with a 6.0.100 SDK installed which is using a 15.4 SDK, causing the app to fall back to `netstandard2.0` or similar (because `16.1` is higher than `15.4`).

It is likely that you are in the second case, and if it is, you're likely building with the `7.0.100` SDK installed and you are then hitting https://github.com/dotnet/sdk/issues/30103. 

There's no fix for it at this time, and there are two known ways to workaround it:
- Build your library for `net6.0-XXX` only (and use a `global.json` to ensure that it is using a `6.0.XXX` SDK), but that will prevent you from adding `net7.0` using standard nuget packaging. Yet, it will allow for `net6.0` apps built with a `6.0.XXX` SDK to use the library.
- Build separately for `net6.0-ios` and `net7.0-ios` and aggregate the results manually with a nuspec file (very error prone and time consuming)

We assume that this will be fixed for .NET 8, but not convinced yet.

**Reply by jeanplevesque** [reply]
Right now we don't specify the iOS SDK version (it's just `net6.0-ios`).
https://github.com/nventive/ExtendedSplashScreen/blob/master/src/library/ExtendedSplashScreen.Uno.WinUI/ExtendedSplashScreen.Uno.WinUI.csproj#L3

**Reply by jeromelaban** [reply]
Then this means that you're building with this particular library with a .NET 7.0.XX SDK installed (even if you're specifying `net6.0-ios` explicitly in the project), and that will fail in a net6.0 built app (if built using .NET 6.0.XXX installed).

You'll need to adjust the CI for [https://github.com/nventive/ExtendedSplashScreen](https://github.com/nventive/ExtendedSplashScreen/blob/master/src/library/ExtendedSplashScreen.Uno.WinUI/ExtendedSplashScreen.Uno.WinUI.csproj#L3) to use .NET 6.0.XXX to build.

**Reply by jeanplevesque** [reply]
So I tried to specify the .NET SDK version and it worked locally, but not with the CI build... I'm not sure why?
https://github.com/nventive/ExtendedSplashScreen/pull/66/files

**Reply by jeromelaban** [reply]
Can you provide a binlog?

**Reply by jeanplevesque** [reply]
[build.binlog.txt](https://github.com/unoplatform/uno/files/10855349/build.binlog.txt)
(Just remove the .txt extension.)

**Reply by jeromelaban** [reply]
I'm not seeing any failure in the binlog, what's the error you're seeing?

**Reply by jeanplevesque** [reply]
I have the same runtime error.
`ObjCRuntime.RuntimeException: 'Can't register the class Nventive.ExtendedSplashScreen.ExtendedSplashScreen when the dynamic registrar has been linked away.'`

**Reply by jeromelaban** [reply]
Your package is still building with the net7 SDK (see the 16.1 at the end):
![image](https://user-images.githubusercontent.com/5839577/223131190-ec99cb95-ce53-4073-8a39-ee88af3d1e63.png)

It may be that running uno-check like in https://github.com/unoplatform/uno.toolkit.ui/pull/488 will help forcing the SDK to be at the right version.

**Reply by jeanplevesque** [reply]
hmm that would mean that the `global.json` is not actually honored 😯

**Reply by jeromelaban** [reply]
It could very well be for workloads, workloads are troublesome. You can see in the binlogs that the proper SDK is used, but workloads are not.
